File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.cpp
class Solution {
public:
    int shortestPath(vector<vector<int>>& grid, int k) {
        
        int n=grid.size();
        int m=grid[0].size();
        
        if(n == 1 && m == 1) return 0;
        
        queue<tuple<int,int,int>> q;
        q.push({0, 0, k});
        
        int dir[4][2] = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        vector<vector<int>> kVis(n, vector<int>(m, -1));
        kVis[0][0] = k;
        
        int level=0;
        while(q.size() > 0){
            
            int size=q.size();
            while(size-- > 0){
                
                auto [curRow, curCol, curK] = q.front();
                q.pop();
                
                for(int i=0; i<4; i++){
                    int nextRow = curRow + dir[i][0];
                    int nextCol = curCol + dir[i][1];
                    
                    if(nextRow == n-1 && nextCol == m-1) return level+1;
                    
                    if(nextRow >= 0 && nextRow < n && nextCol >= 0 && nextCol < m){
                        
                        if(grid[nextRow][nextCol] == 1 && curK>0 && kVis[nextRow][nextCol] < (curK - 1)){
                            q.push({nextRow, nextCol, curK-1});
                            kVis[nextRow][nextCol] = curK - 1;
                        }
                        
                        else if(grid[nextRow][nextCol] == 0 && kVis[nextRow][nextCol] < (curK)){
                            q.push({nextRow, nextCol, curK});
                            kVis[nextRow][nextCol] = curK;
                        }
                    }
                }
            }
            level++;
        }
        
        return -1;
    }
};

File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.java
class Node{
    int r;
    int c;
    int k;
    
    Node(int r, int c, int k){
        this.r = r;
        this.c = c;
        this.k = k;
    }
}

class Solution {
    public int shortestPath(int[][] grid, int k) {
        
        int ROWS = grid.length, COLS = grid[0].length;
        
        if(ROWS == 1 && COLS == 1) return 0;
        
        int[][] dir = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        
        int[][] kVis = new int[ROWS][COLS];
        
        for(int[] row : kVis) Arrays.fill(row, -1);
        kVis[0][0] = k;
        
        Queue<Node> q = new LinkedList<>();
        q.add(new Node(0, 0, k));
        
        int level=0;
        
        while(q.size() > 0){
            
            int size = q.size();
            while(size-- > 0){
                
                Node curNode = q.poll();
                
                for(int i=0; i<dir.length; i++){
                    int nextRow = curNode.r + dir[i][0];
                    int nextCol = curNode.c + dir[i][1];
                    
                    if(nextRow == ROWS-1 && nextCol == COLS-1) return level+1;
                    
                    if(nextRow >= 0 && nextRow < ROWS && nextCol >= 0 && nextCol < COLS){
                        
                        if(grid[nextRow][nextCol] == 1 && curNode.k > 0 && kVis[nextRow][nextCol] < curNode.k - 1){
                            q.add(new Node(nextRow, nextCol, curNode.k - 1));
                            kVis[nextRow][nextCol] = curNode.k - 1;
                        }
                        else if(grid[nextRow][nextCol] == 0 && kVis[nextRow][nextCol] < curNode.k){
                            q.add(new Node(nextRow, nextCol, curNode.k));
                            kVis[nextRow][nextCol] = curNode.k;
                        }
                    }
                }
            }
            level++;
        }
        
        return -1;
        
    }
}